game-core/src/main/java/games/strategy/engine/framework/startup/launcher/AbstractLauncher.java

```diff
@@ -2,52 +2,20 @@
 
 import java.awt.Component;
 
-import javax.swing.JOptionPane;
-import javax.swing.SwingUtilities;
-
-import games.strategy.engine.data.GameData;
-import games.strategy.engine.framework.startup.mc.GameSelectorModel;
-import games.strategy.engine.framework.ui.background.WaitWindow;
-
 /**
  * Abstract class for launching a game.
+ *
+ * @param <T> The type of object that gets returned by {@link #loadGame(Component)}
+ *        and is required by {@link #launchInternal(Component, Object)}.
  */
-public abstract class AbstractLauncher implements ILauncher {
-  protected final GameData gameData;
-  protected final GameSelectorModel gameSelectorModel;
-  protected final WaitWindow gameLoadingWindow;
-  protected final boolean headless;
-
-  protected AbstractLauncher(final GameSelectorModel gameSelectorModel) {
-    this(gameSelectorModel, false);
-  }
-
-  protected AbstractLauncher(final GameSelectorModel gameSelectorModel, final boolean headless) {
-    this.headless = headless;
-    if (this.headless) {
-      gameLoadingWindow = null;
-    } else {
-      gameLoadingWindow = new WaitWindow();
-    }
-    this.gameSelectorModel = gameSelectorModel;
-    gameData = gameSelectorModel.getGameData();
-  }
-
+public abstract class AbstractLauncher<T> implements ILauncher {
   @Override
   public void launch(final Component parent) {
-    if (!headless && !SwingUtilities.isEventDispatchThread()) {
-      throw new IllegalStateException("Wrong thread");
-    }
-    if (!headless && gameLoadingWindow != null) {
-      gameLoadingWindow.setLocationRelativeTo(JOptionPane.getFrameForComponent(parent));
-      gameLoadingWindow.setVisible(true);
-      gameLoadingWindow.showWait();
-    }
-    if (parent != null) {
-      JOptionPane.getFrameForComponent(parent).setVisible(false);
-    }
-    new Thread(() -> launchInNewThread(parent), "Triplea start thread").start();
+    final T result = loadGame(parent);
+    new Thread(() -> launchInternal(parent, result)).start();
   }
 
-  protected abstract void launchInNewThread(Component parent);
+  abstract T loadGame(Component parent);
```

Comment:
Any thoughts if it would simplify to use `Optional<T>` instead of `T`?

For example, could write something like:
```
loadGame(parent).ifPresent(
  () ->  new Thread(
          () -> launchInternal(parent, result)).start());
```

Reply:
Hmm doesn't really seem to work with the current launch structure. There are some finally blocks that must be executed anyways.
I tried using it anyways. Let me know if that's ok for you.